File: src/orient/repo_index.py
from __future__ import annotations

import ast
import math
import re
from collections import Counter
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable


IGNORED_DIRS = {
    ".git",
    ".venv",
    "__pycache__",
    ".pytest_cache",
    "node_modules",
    "dist",
    "build",
    ".next",
    "coverage",
}

EXT_LANG = {
    ".py": "python",
    ".js": "javascript",
    ".jsx": "javascript",
    ".ts": "typescript",
    ".tsx": "typescript",
    ".go": "go",
    ".rs": "rust",
    ".rb": "ruby",
    ".java": "java",
    ".kt": "kotlin",
    ".swift": "swift",
    ".md": "markdown",
    ".toml": "toml",
    ".json": "json",
    ".yaml": "yaml",
    ".yml": "yaml",
}


@dataclass(frozen=True)
class Symbol:
    name: str
    kind: str
    path: str
    line: int


@dataclass(frozen=True)
class SearchResult:
    path: str
    score: float
    reason: str
    snippet: str


@dataclass(frozen=True)
class RelatedFile:
    path: str
    reason: str
    score: float


@dataclass(frozen=True)
class RepoBrief:
    root_name: str
    file_count: int
    language_counts: dict[str, int]
    known_commands: list[str]
    important_files: list[str]


@dataclass
class IndexedFile:
    path: str
    language: str
    text: str
    tokens: Counter[str] = field(default_factory=Counter)
    symbols: list[Symbol] = field(default_factory=list)


class RepoIndex:
    def __init__(self, root: Path, files: dict[str, IndexedFile]):
        self.root = root
        self.files = files
        self.symbols = [symbol for file in files.values() for symbol in file.symbols]
        self._doc_freq = self._build_doc_freq()

    def find_symbol(self, name: str, limit: int = 10) -> list[Symbol]:
        needle = normalize_token(name)
        scored: list[tuple[int, Symbol]] = []
        for symbol in self.symbols:
            symbol_token = normalize_token(symbol.name)
            if symbol.name == name:
                scored.append((100, symbol))
            elif symbol_token == needle:
                scored.append((90, symbol))
            elif needle in symbol_token:
                scored.append((60, symbol))
        return [symbol for _, symbol in sorted(scored, key=lambda item: (-item[0], item[1].path, item[1].line))[:limit]]

    def search_code(self, query: str, limit: int = 10) -> list[SearchResult]:
        query_tokens = tokenize(query)
        if not query_tokens:
            return []
        results: list[SearchResult] = []
        total_docs = max(len(self.files), 1)
        for file in self.files.values():
            score = 0.0
            reasons: list[str] = []
            for token in query_tokens:
                tf = file.tokens.get(token, 0)
                if not tf:
                    continue
                idf = math.log((total_docs + 1) / (self._doc_freq[token] + 1)) + 1
                score += (1 + math.log(tf)) * idf
                reasons.append(token)
            for symbol in file.symbols:
                symbol_tokens = tokenize(symbol.name)
                overlap = set(symbol_tokens) & set(query_tokens)
                if overlap:
                    score += 2.0 * len(overlap)
                    reasons.extend(sorted(overlap))
            if score:
                results.append(
                    SearchResult(
                        path=file.path,
                        score=round(score, 4),
                        reason="matched " + ", ".join(sorted(set(reasons))[:8]),
                        snippet=best_snippet(file.text, query_tokens),
                    )
                )
        return sorted(results, key=lambda item: (-item.score, item.path))[:limit]

    def related_files(self, path: str, limit: int = 10) -> list[RelatedFile]:
        normalized = path.strip("/")
        stem = Path(normalized).stem
        directory = str(Path(normalized).parent)
        related: list[RelatedFile] = []
        for file_path in self.files:
            if file_path == normalized:
                continue
            score = 0.0
            reasons: list[str] = []
            lower = file_path.lower()
            if stem and stem.lower() in lower:
                score += 4
                reasons.append(f"shares stem {stem}")
            if lower.startswith("test") or "/test" in lower or lower.endswith("_test.py") or lower.endswith(".test.ts"):
                source_stem = stem.replace("test_", "")
                if source_stem and source_stem.lower() in lower:
                    score += 5
                    reasons.append("test coverage candidate")
            if directory != "." and str(Path(file_path).parent) == directory:
                score += 1
                reasons.append("same directory")
            if score:
                related.append(RelatedFile(file_path, "; ".join(reasons), score))
        return sorted(related, key=lambda item: (-item.score, item.path))[:limit]

    def repo_brief(self) -> RepoBrief:
        language_counts = Counter(file.language for file in self.files.values())
        important = [
            name
            for name in ["AGENTS.md", "CLAUDE.md", "README.md", "pyproject.toml", "package.json", "Makefile"]
            if name in self.files
        ]
        return RepoBrief(
            root_name=self.root.name,
            file_count=len(self.files),
            language_counts=dict(language_counts),
            known_commands=known_commands(self.root, self.files),
            important_files=important,
        )

    def _build_doc_freq(self) -> Counter[str]:
        counts: Counter[str] = Counter()
        for file in self.files.values():
            counts.update(file.tokens.keys())
        return counts


class RepoIndexer:
    def __init__(self, root: Path | str):
        self.root = Path(root).resolve()

    def build(self) -> RepoIndex:
        files: dict[str, IndexedFile] = {}
        for path in walk_files(self.root):
            rel = path.relative_to(self.root).as_posix()
            language = EXT_LANG.get(path.suffix.lower(), "text")
            try:
                text = path.read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue
            if "\0" in text:
                continue
            symbols = extract_symbols(rel, text, language)
            files[rel] = IndexedFile(rel, language, text, Counter(tokenize(text + " " + rel)), symbols)
        return RepoIndex(self.root, files)


def walk_files(root: Path) -> Iterable[Path]:
    for path in root.rglob("*"):
        if path.is_dir():
            continue
        if any(part in IGNORED_DIRS for part in path.relative_to(root).parts):
            continue
        if path.stat().st_size > 512_000:
            continue
        if path.suffix.lower() not in EXT_LANG and path.name not in {"README", "Makefile", "AGENTS.md", "CLAUDE.md"}:
            continue
        yield path


def extract_symbols(path: str, text: str, language: str) -> list[Symbol]:
    if language == "python":
        return extract_python_symbols(path, text)
    symbols: list[Symbol] = []
    for line_number, line in enumerate(text.splitlines(), start=1):
        match = re.search(r"\b(?:function|class|interface|const|let|var)\s+([A-Za-z_$][\w$]*)", line)
        if match:
            kind = "class" if "class " in line else "function"
            symbols.append(Symbol(match.group(1), kind, path, line_number))
    return symbols


def extract_python_symbols(path: str, text: str) -> list[Symbol]:
    try:
        tree = ast.parse(text)
    except SyntaxError:
        return []
    symbols: list[Symbol] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.ClassDef):
            symbols.append(Symbol(node.name, "class", path, node.lineno))
        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            symbols.append(Symbol(node.name, "function", path, node.lineno))
    return symbols


def known_commands(root: Path, files: dict[str, IndexedFile]) -> list[str]:
    commands: list[str] = []
    if "pyproject.toml" in files:
        commands.append("pytest")
    if "package.json" in files:
        commands.extend(["npm test", "npm run lint"])
    if "Makefile" in files:
        commands.append("make test")
    return commands


def tokenize(text: str) -> list[str]:
    return [token for token in re.findall(r"[A-Za-z][A-Za-z0-9_]*", split_camel(text).lower()) if len(token) > 1]


def split_camel(text: str) -> str:
    return re.sub(r"([a-z0-9])([A-Z])", r"\1 \2", text)


def normalize_token(text: str) -> str:
    return "".join(tokenize(text))


def best_snippet(text: str, query_tokens: list[str]) -> str:
    lines = text.splitlines()
    if not lines:
        return ""
    for idx, line in enumerate(lines):
        lowered = line.lower()
        if any(token in lowered for token in query_tokens):
            start = max(0, idx - 1)
            end = min(len(lines), idx + 3)
            return "\n".join(lines[start:end])[:700]
    return "\n".join(lines[:6])[:700]

File: src-rs/repo_index.rs
//! Repo orientation index.

use anyhow::Result;
use ignore::WalkBuilder;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};

const MAX_FILE_BYTES: u64 = 512_000;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Symbol {
    pub name: String,
    pub kind: String,
    pub path: String,
    pub line: usize,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SearchResult {
    pub path: String,
    pub score: f64,
    pub reason: String,
    pub snippet: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RelatedFile {
    pub path: String,
    pub reason: String,
    pub score: f64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepoBrief {
    pub root_name: String,
    pub file_count: usize,
    pub language_counts: HashMap<String, usize>,
    pub known_commands: Vec<String>,
    pub important_files: Vec<String>,
}

#[derive(Debug, Clone)]
struct IndexedFile {
    path: String,
    language: String,
    text: String,
    tokens: HashMap<String, usize>,
    symbols: Vec<Symbol>,
}

#[derive(Debug, Clone)]
pub struct RepoIndex {
    root: PathBuf,
    files: HashMap<String, IndexedFile>,
    symbols: Vec<Symbol>,
    doc_freq: HashMap<String, usize>,
}

#[derive(Debug, Clone)]
pub struct RepoIndexer {
    root: PathBuf,
}

impl RepoIndexer {
    pub fn new(root: impl AsRef<Path>) -> Self {
        Self {
            root: root.as_ref().to_path_buf(),
        }
    }

    pub fn build(&self) -> Result<RepoIndex> {
        let root = self.root.canonicalize()?;
        let mut files = HashMap::new();

        for entry in WalkBuilder::new(&root)
            .hidden(false)
            .filter_entry(|entry| !is_ignored(entry.path()))
            .build()
        {
            let entry = entry?;
            let path = entry.path();
            if !path.is_file() {
                continue;
            }
            let metadata = entry.metadata()?;
            if metadata.len() > MAX_FILE_BYTES {
                continue;
            }
            let Some(language) = language_for(path) else {
                continue;
            };
            let text = fs::read_to_string(path).unwrap_or_default();
            if text.contains('\0') {
                continue;
            }
            let rel = path.strip_prefix(&root)?.to_string_lossy().replace('\\', "/");
            let symbols = extract_symbols(&rel, &text, &language);
            let tokens = token_counts(&format!("{rel}\n{text}"));
            files.insert(
                rel.clone(),
                IndexedFile {
                    path: rel,
                    language,
                    text,
                    tokens,
                    symbols,
                },
            );
        }

        let symbols = files
            .values()
            .flat_map(|file| file.symbols.clone())
            .collect::<Vec<_>>();
        let doc_freq = build_doc_freq(&files);

        Ok(RepoIndex {
            root,
            files,
            symbols,
            doc_freq,
        })
    }
}

impl RepoIndex {
    pub fn find_symbol(&self, name: &str, limit: usize) -> Vec<Symbol> {
        let needle = normalize_token(name);
        let mut scored = Vec::new();
        for symbol in &self.symbols {
            let symbol_token = normalize_token(&symbol.name);
            let score = if symbol.name == name {
                100
            } else if symbol_token == needle {
                90
            } else if symbol_token.contains(&needle) {
                60
            } else {
                0
            };
            if score > 0 {
                scored.push((score, symbol.clone()));
            }
        }
        scored.sort_by(|a, b| b.0.cmp(&a.0).then_with(|| a.1.path.cmp(&b.1.path)));
        scored
            .into_iter()
            .take(limit)
            .map(|(_, symbol)| symbol)
            .collect()
    }

    pub fn search_code(&self, query: &str, limit: usize) -> Vec<SearchResult> {
        let query_tokens = tokenize(query);
        if query_tokens.is_empty() {
            return Vec::new();
        }
        let query_set = query_tokens.iter().cloned().collect::<HashSet<_>>();
        let total_docs = self.files.len().max(1) as f64;
        let mut results = Vec::new();

        for file in self.files.values() {
            let mut score = 0.0;
            let mut reasons = HashSet::new();
            for token in &query_tokens {
                let Some(tf) = file.tokens.get(token) else {
                    continue;
                };
                let df = *self.doc_freq.get(token).unwrap_or(&0) as f64;
                let idf = ((total_docs + 1.0) / (df + 1.0)).ln() + 1.0;
                score += (1.0 + (*tf as f64).ln()) * idf;
                reasons.insert(token.clone());
            }
            for symbol in &file.symbols {
                let overlap = tokenize(&symbol.name)
                    .into_iter()
                    .filter(|token| query_set.contains(token))
                    .collect::<Vec<_>>();
                if !overlap.is_empty() {
                    score += 2.0 * overlap.len() as f64;
                    for token in overlap {
                        reasons.insert(token);
                    }
                }
            }
            if score > 0.0 {
                let mut reasons = reasons.into_iter().collect::<Vec<_>>();
                reasons.sort();
                results.push(SearchResult {
                    path: file.path.clone(),
                    score: round4(score),
                    reason: format!("matched {}", reasons.join(", ")),
                    snippet: best_snippet(&file.text, &query_tokens),
                });
            }
        }

        results.sort_by(|a, b| {
            b.score
                .partial_cmp(&a.score)
                .unwrap_or(Ordering::Equal)
                .then_with(|| a.path.cmp(&b.path))
        });
        results.truncate(limit);
        results
    }

    pub fn related_files(&self, path: &str, limit: usize) -> Vec<RelatedFile> {
        let normalized = path.trim_start_matches('/').to_string();
        let stem = Path::new(&normalized)
            .file_stem()
            .map(|value| value.to_string_lossy().to_string())
            .unwrap_or_default();
        let directory = Path::new(&normalized)
            .parent()
            .map(|value| value.to_string_lossy().to_string())
            .unwrap_or_default();
        let mut related = Vec::new();
        for file_path in self.files.keys() {
            if file_path == &normalized {
                continue;
            }
            let lower = file_path.to_lowercase();
            let mut score = 0.0;
            let mut reasons = Vec::new();
            if !stem.is_empty() && lower.contains(&stem.to_lowercase()) {
                score += 4.0;
                reasons.push(format!("shares stem {stem}"));
            }
            if is_test_path(&lower) && !stem.is_empty() && lower.contains(&stem.replace("test_", "").to_lowercase()) {
                score += 5.0;
                reasons.push("test coverage candidate".to_string());
            }
            let file_dir = Path::new(file_path)
                .parent()
                .map(|value| value.to_string_lossy().to_string())
                .unwrap_or_default();
            if !directory.is_empty() && file_dir == directory {
                score += 1.0;
                reasons.push("same directory".to_string());
            }
            if score > 0.0 {
                related.push(RelatedFile {
                    path: file_path.clone(),
                    reason: reasons.join("; "),
                    score,
                });
            }
        }
        related.sort_by(|a, b| {
            b.score
                .partial_cmp(&a.score)
                .unwrap_or(Ordering::Equal)
                .then_with(|| a.path.cmp(&b.path))
        });
        related.truncate(limit);
        related
    }

    pub fn repo_brief(&self) -> RepoBrief {
        let mut language_counts = HashMap::new();
        for file in self.files.values() {
            *language_counts.entry(file.language.clone()).or_insert(0) += 1;
        }
        let important_files = [
            "AGENTS.md",
            "CLAUDE.md",
            "README.md",
            "pyproject.toml",
            "package.json",
            "Cargo.toml",
            "Makefile",
        ]
        .into_iter()
        .filter(|name| self.files.contains_key(*name))
        .map(String::from)
        .collect();

        RepoBrief {
            root_name: self
                .root
                .file_name()
                .map(|value| value.to_string_lossy().to_string())
                .unwrap_or_else(|| self.root.display().to_string()),
            file_count: self.files.len(),
            language_counts,
            known_commands: self.known_commands(),
            important_files,
        }
    }

    fn known_commands(&self) -> Vec<String> {
        let mut commands = Vec::new();
        if self.files.contains_key("pyproject.toml") {
            commands.push("pytest".to_string());
        }
        if self.files.contains_key("Cargo.toml") {
            commands.push("cargo test".to_string());
        }
        if self.files.contains_key("package.json") {
            commands.push("npm test".to_string());
            commands.push("npm run lint".to_string());
        }
        if self.files.contains_key("Makefile") {
            commands.push("make test".to_string());
        }
        commands
    }
}

fn is_ignored(path: &Path) -> bool {
    path.components().any(|component| {
        let part = component.as_os_str().to_string_lossy();
        matches!(
            part.as_ref(),
            ".git"
                | ".venv"
                | "__pycache__"
                | ".pytest_cache"
                | "node_modules"
                | "dist"
                | "build"
                | ".next"
                | "coverage"
                | "target"
        )
    })
}

fn language_for(path: &Path) -> Option<String> {
    let file_name = path.file_name()?.to_string_lossy();
    if matches!(file_name.as_ref(), "README" | "README.md" | "AGENTS.md" | "CLAUDE.md" | "Makefile") {
        return Some("text".to_string());
    }
    let ext = path.extension()?.to_string_lossy().to_lowercase();
    let language = match ext.as_str() {
        "py" => "python",
        "rs" => "rust",
        "js" | "jsx" => "javascript",
        "ts" | "tsx" => "typescript",
        "go" => "go",
        "rb" => "ruby",
        "java" => "java",
        "kt" => "kotlin",
        "swift" => "swift",
        "md" => "markdown",
        "toml" => "toml",
        "json" => "json",
        "yaml" | "yml" => "yaml",
        _ => return None,
    };
    Some(language.to_string())
}

fn extract_symbols(path: &str, text: &str, language: &str) -> Vec<Symbol> {
    if language == "python" {
        return extract_python_symbols(path, text);
    }
    let re = Regex::new(r"\b(?:pub\s+)?(?:async\s+)?(?:fn|function|class|interface|struct|enum|const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)").unwrap();
    text.lines()
        .enumerate()
        .filter_map(|(index, line)| {
            let capture = re.captures(line)?;
            let kind = if line.contains("class ") {
                "class"
            } else if line.contains("struct ") {
                "struct"
            } else if line.contains("enum ") {
                "enum"
            } else {
                "function"
            };
            Some(Symbol {
                name: capture.get(1)?.as_str().to_string(),
                kind: kind.to_string(),
                path: path.to_string(),
                line: index + 1,
            })
        })
        .collect()
}

fn extract_python_symbols(path: &str, text: &str) -> Vec<Symbol> {
    let re = Regex::new(r"^\s*(class|def|async\s+def)\s+([A-Za-z_][A-Za-z0-9_]*)").unwrap();
    text.lines()
        .enumerate()
        .filter_map(|(index, line)| {
            let capture = re.captures(line)?;
            let raw_kind = capture.get(1)?.as_str();
            Some(Symbol {
                name: capture.get(2)?.as_str().to_string(),
                kind: if raw_kind == "class" { "class" } else { "function" }.to_string(),
                path: path.to_string(),
                line: index + 1,
            })
        })
        .collect()
}

fn build_doc_freq(files: &HashMap<String, IndexedFile>) -> HashMap<String, usize> {
    let mut doc_freq = HashMap::new();
    for file in files.values() {
        for token in file.tokens.keys() {
            *doc_freq.entry(token.clone()).or_insert(0) += 1;
        }
    }
    doc_freq
}

fn token_counts(text: &str) -> HashMap<String, usize> {
    let mut counts = HashMap::new();
    for token in tokenize(text) {
        *counts.entry(token).or_insert(0) += 1;
    }
    counts
}

fn tokenize(text: &str) -> Vec<String> {
    let camel = Regex::new(r"([a-z0-9])([A-Z])").unwrap();
    let split = camel.replace_all(text, "$1 $2");
    let re = Regex::new(r"[A-Za-z][A-Za-z0-9_]*").unwrap();
    re.find_iter(&split)
        .map(|m| m.as_str().to_lowercase())
        .filter(|token| token.len() > 1)
        .collect()
}

fn normalize_token(text: &str) -> String {
    tokenize(text).join("")
}

fn best_snippet(text: &str, query_tokens: &[String]) -> String {
    let lines = text.lines().collect::<Vec<_>>();
    for (idx, line) in lines.iter().enumerate() {
        let lowered = line.to_lowercase();
        if query_tokens.iter().any(|token| lowered.contains(token)) {
            let start = idx.saturating_sub(1);
            let end = (idx + 3).min(lines.len());
            return lines[start..end].join("\n").chars().take(700).collect();
        }
    }
    lines.into_iter().take(6).collect::<Vec<_>>().join("\n")
}

fn is_test_path(path: &str) -> bool {
    path.starts_with("test")
        || path.contains("/test")
        || path.ends_with("_test.py")
        || path.ends_with(".test.ts")
        || path.ends_with(".test.tsx")
}

fn round4(value: f64) -> f64 {
    (value * 10_000.0).round() / 10_000.0
}
